File: pySSS/resources/SubtitleCorpusReader.py
from os import walk
from dialog import SimpleQA
from org.apache.lucene.document import Document, TextField, Field

import transaction

def read(writer, dbroot, corpusDir):  #missing arg: normalizers
    files = []

    #extract filenames from corpus directory
    for (dirpath, dirnames, filenames) in walk(corpusDir):
        files.extend(filenames)
        break

    i = 0

    for filename in files:
        print("Creating lucene indexes and database for", filename, "...")
        filepath = corpusDir+'/'+filename
        file = open(filepath, 'r')

        lineNum = 0
        previousDialogId = 0
        internalId = -1

        line = file.readline()

        while (line):
            lineNum += 1

            if not line.strip():        #skip empty lines
                line = file.readline()
                continue;

            temp = line
            assert temp.startswith("SubId")
            subId = getSubstringAfterHyphen(temp)

            temp = file.readline()
            lineNum += 1
            assert temp.startswith("DialogId")
            dialogId = int(getSubstringAfterHyphen(temp))

            temp = file.readline()
            lineNum += 1
            assert temp.startswith("Diff")
            diff = float(getSubstringAfterHyphen(temp))

            temp = file.readline()
            lineNum += 1
            assert temp.startswith("I")
            question = getSubstringAfterHyphen(temp)

            temp = file.readline()
            lineNum += 1
            assert temp.startswith("R")
            answer = getSubstringAfterHyphen(temp)

            answer = answer.strip()

            normalizedAnswer = normalize(answer)
            normalizedQuestion = normalize(question)

            #apply normalization to answer and question
            if dialogId == previousDialogId + 1:
                simpleQA = SimpleQA.SimpleQA(internalId, question, normalizedQuestion, answer, normalizedAnswer, diff)
            else:
                simpleQA = SimpleQA.SimpleQA(-1, question, normalizedQuestion, answer, normalizedAnswer, diff)

            dbroot['simpleQAs'][lineNum] = simpleQA    #db.store(simpleQA);

            del simpleQA
            previousDialogId = dialogId

            addDoc(writer, question, str(lineNum)) #using lineNum as id

            line = file.readline()

        #print("final:", lineNum)
        i += 1

        print()

        if i % 200 == 0:
            transaction.commit()

    transaction.commit()
        #for sqa in dbroot['simpleQAs'].values():
        #    print(sqa.getDiff())

def getSubstringAfterHyphen(temp):
    return temp[temp.index('-') + 2:]

def addDoc(writer, question, answer):
    doc = Document()
    doc.add(TextField('question', question, Field.Store.YES))
    doc.add(TextField('answer', answer, Field.Store.YES))
    writer.addDocument(doc)

def normalize(query):
    query_normalized = query.lower()
    puncts = ['"','.','?','!',',', '\n']
    for sym in puncts:
        query_normalized = query_normalized.replace(sym, '')
    return query_normalized

File: pySSS/lucene_resources/SubtitleCorpusReader.java
package lucene_resources;

import com.db4o.ObjectContainer;
import org.apache.lucene.index.IndexWriter;
//import SimpleQA;
//import sss.texttools.Cleaner;
//import sss.texttools.normalizer.Normalizer;

import java.io.*;
import java.text.NumberFormat;
import java.util.List;
import java.util.Arrays;

public class SubtitleCorpusReader extends CorpusReader {


    @Override
    public void read(IndexWriter writer, ObjectContainer db, File[] files) throws IOException {

//        Cleaner c = new Cleaner(); // Marco

        int i = 0;

        for (File file : files) {

            System.out.println("Creating lucene indexes and database for " + file.getName() + "...");
            BufferedReader reader = new BufferedReader(new FileReader(file.getCanonicalPath()));

            String line;

            String subId;
            String question;
            String answer;

            long internalId = -1;
            int previousDialogId = 0;
            long totalLines = count(file.getCanonicalPath());
            long step = totalLines / 1000;
            long lineNum = 0;

            // Marco
//            System.out.println(((double)(i*100) / files.length));

            while ((line = reader.readLine()) != null) {

                lineNum++;
                //if ((lineNum % step) == 0) {
                //    System.out.println(getPercentage(lineNum, totalLines));
                //}

                if (line.trim().length() == 0) {
                    continue;
                }

                String temp = line;
                assert (temp.startsWith("SubId"));
                subId = getSubstringAfterHyphen(temp);

                temp = reader.readLine(); lineNum++;
                assert (temp.startsWith("DialogId"));
                int dialogId = Integer.parseInt(getSubstringAfterHyphen(temp));

                temp = reader.readLine(); lineNum++;
                assert (temp.startsWith("Diff"));
                long diff = Long.parseLong(getSubstringAfterHyphen(temp));

                temp = reader.readLine(); lineNum++;
                assert (temp.startsWith("I"));
                question = getSubstringAfterHyphen(temp); //assumes the corpus does not have empty questions

                temp = reader.readLine(); lineNum++;
                assert (temp.startsWith("R"));
                answer = getSubstringAfterHyphen(temp); //assumes the corpus does not have empty answers

                answer = answer.trim();

                String normalizedAnswer = normalize(answer);
                String normalizedQuestion = normalize(question);
                // Marco
//                String normalizedAnswer = c.cleanAndNormalize(answer);
//                String normalizedQuestion = c.cleanAndNormalize(question);

                SimpleQA simpleQA;

                if (dialogId == previousDialogId + 1) {
                    simpleQA = new SimpleQA(internalId, question, answer, normalizedQuestion, normalizedAnswer, diff);
                } else {
                    simpleQA = new SimpleQA(-1, question, answer, normalizedQuestion, normalizedAnswer, diff);
                }

                db.store(simpleQA);
                internalId = db.ext().getID(simpleQA);
                simpleQA = null;
                previousDialogId = dialogId;
                addDoc(writer, normalizedQuestion, String.valueOf(internalId));
            }

            i++;
            System.out.println();

            // Marco
            if (i % 200 == 0) {
                db.commit();
            }
        }

        db.commit();
    }

    private String normalize(String query){
        String query_normalized = query.toLowerCase();
        List<String> puncts = Arrays.asList(".","?","!",",", "\n", "-");
        for (String sym : puncts){
            query_normalized = query_normalized.replace(sym, "");
        }
        return query_normalized;
    }


    private String getPercentage(long partial, long total) {
        NumberFormat defaultFormat = NumberFormat.getPercentInstance();
        defaultFormat.setMinimumFractionDigits(1);
        return defaultFormat.format(partial / (double) total);
    }

    private String getSubstringAfterHyphen(String temp) {
        return temp.substring(temp.indexOf('-') + 2, temp.length());
    }

    private int count(String filename) throws IOException {
        InputStream is = new BufferedInputStream(new FileInputStream(filename));
        try {
            byte[] c = new byte[1024];
            int count = 0;
            int readChars = 0;
            boolean endsWithoutNewLine = false;
            while ((readChars = is.read(c)) != -1) {
                for (int i = 0; i < readChars; ++i) {
                    if (c[i] == '\n')
                        ++count;
                }
                endsWithoutNewLine = (c[readChars - 1] != '\n');
            }
            if(endsWithoutNewLine) {
                ++count;
            }
            return count;
        } finally {
            is.close();
        }
    }
}
